module/zfs/zstd.c

```diff
@@ -0,0 +1,617 @@
+			/* Iterate rather than recurse to save stack space */
+			/* qsort(pn - d2, d2 / es, es, cmp); */
+			a = pn - d2;
+			n = d2 / es;
+			goto loop;
+		}
+	} else {
+		/* Recurse on right partition, then iterate on left partition */
+		if (d2 > es) {
+			zstd_qsort(pn - d2, d2 / es, es, cmp);
+		}
+		if (d1 > es) {
+			/* Iterate rather than recurse to save stack space */
+			/* qsort(a, d1 / es, es, cmp); */
+			n = d1 / es;
+			goto loop;
+		}
+	}
+}
+
+
+static size_t real_zstd_compress(const char *source, char *dest, int isize,
+    int osize, int level);
+static size_t real_zstd_decompress(const char *source, char *dest, int isize,
+    int maxosize);
+
+void *zstd_alloc(void *opaque, size_t size);
+void zstd_free(void *opaque, void *ptr);
+
+static const ZSTD_customMem zstd_malloc = {
+	zstd_alloc,
+	zstd_free,
+	NULL,
+};
+
+enum zstd_kmem_type {
+	ZSTD_KMEM_UNKNOWN = 0,
+	ZSTD_KMEM_CCTX,
+	ZSTD_KMEM_WRKSPC_4K_MIN,
+	ZSTD_KMEM_WRKSPC_4K_DEF,
+	ZSTD_KMEM_WRKSPC_4K_MAX,
+	ZSTD_KMEM_WRKSPC_16K_MIN,
+	ZSTD_KMEM_WRKSPC_16K_DEF,
+	ZSTD_KMEM_WRKSPC_16K_MAX,
+	ZSTD_KMEM_WRKSPC_128K_MIN,
+	ZSTD_KMEM_WRKSPC_128K_DEF,
+	ZSTD_KMEM_WRKSPC_128K_MAX,
+	/* SPA_MAXBLOCKSIZE */
+	ZSTD_KMEM_WRKSPC_MBS_MIN,
+	ZSTD_KMEM_WRKSPC_MBS_DEF,
+	ZSTD_KMEM_WRKSPC_MBS_MAX,
+	ZSTD_KMEM_DCTX,
+	ZSTD_KMEM_COUNT,
+};
+
+struct zstd_kmem {
+	uint_t			kmem_magic;
+	enum zstd_kmem_type	kmem_type;
+	size_t			kmem_size;
+};
+
+struct zstd_kmem_config {
+	size_t			block_size;
+	int			compress_level;
+	char			*cache_name;
+};
+
+static kmem_cache_t *zstd_kmem_cache[ZSTD_KMEM_COUNT] = { NULL };
+static struct zstd_kmem zstd_cache_size[ZSTD_KMEM_COUNT] = {
+	{ ZSTD_KMEM_MAGIC, 0, 0 } };
+static struct zstd_kmem_config zstd_cache_config[ZSTD_KMEM_COUNT] = {
+	{ 0, 0, "zstd_unknown" },
+	{ 0, 0, "zstd_cctx" },
+	{ 4096, ZIO_ZSTD_LEVEL_MIN, "zstd_wrkspc_4k_min" },
+	{ 4096, ZIO_ZSTD_LEVEL_DEFAULT, "zstd_wrkspc_4k_def" },
+	{ 4096, ZIO_ZSTD_LEVEL_MAX, "zstd_wrkspc_4k_max" },
+	{ 16384, ZIO_ZSTD_LEVEL_MIN, "zstd_wrkspc_16k_min" },
+	{ 16384, ZIO_ZSTD_LEVEL_DEFAULT, "zstd_wrkspc_16k_def" },
+	{ 16384, ZIO_ZSTD_LEVEL_MAX, "zstd_wrkspc_16k_max" },
+	{ SPA_OLD_MAXBLOCKSIZE, ZIO_ZSTD_LEVEL_MIN, "zstd_wrkspc_128k_min" },
+	{ SPA_OLD_MAXBLOCKSIZE, ZIO_ZSTD_LEVEL_DEFAULT,
+	    "zstd_wrkspc_128k_def" },
+	{ SPA_OLD_MAXBLOCKSIZE, ZIO_ZSTD_LEVEL_MAX, "zstd_wrkspc_128k_max" },
+	{ SPA_MAXBLOCKSIZE, ZIO_ZSTD_LEVEL_MIN, "zstd_wrkspc_mbs_min" },
+	{ SPA_MAXBLOCKSIZE, ZIO_ZSTD_LEVEL_DEFAULT, "zstd_wrkspc_mbs_def" },
+	{ SPA_MAXBLOCKSIZE, ZIO_ZSTD_LEVEL_MAX, "zstd_wrkspc_mbs_max" },
+	{ 0, 0, "zstd_dctx" },
+};
+
+static int
+zstd_compare(const void *a, const void *b)
+{
+	struct zstd_kmem *x, *y;
+
+	x = (struct zstd_kmem *)a;
+	y = (struct zstd_kmem *)b;
+
+	ASSERT(x->kmem_magic == ZSTD_KMEM_MAGIC);
+	ASSERT(y->kmem_magic == ZSTD_KMEM_MAGIC);
+
+	if (x->kmem_size > y->kmem_size) {
+		return (1);
+	} else if (x->kmem_size == y->kmem_size) {
+		return (0);
+	} else {
+		return (-1);
+	}
```

Comment:
You can use `AVL_CMP` here which is significantly faster because it requires no branching.